CMakeLists.txt

```diff
@@ -306,6 +306,15 @@ else()
     endif()
 endif()
 
+if(NOT WIN32 AND NOT APPLE)
+  find_library(LIBMLIRMIOPEN MLIRMIOpen)
+  if(LIBMLIRMIOPEN AND NOT MIOPEN_USE_COMGR)
```

Comment:
Why COMgr can't be used together with MLIR?

Anyway, please do like this:
```
    if(MIOPEN_USE_COMGR)
        message(FATAL_ERROR "comgr cannot be used with MLIRMIOpen")
    endif()
```

Reply:
There is a CI pipeline stage that specifically test COMGR. If I do as suggested, it will break that stage.

Back to the original question, COMgr can use together with MLIR. It is just additional development/plumbing effort to aggregate the MLIR build logic to COMgr path. According to my 1:1 discussion with @daniellowell, COMgr is targeted for a very specific deliverable and can be skipped in the MLIR development. If that is not the case now, let me know. Thanks.

Reply:
> There is a CI pipeline stage that specifically test COMGR. If I do as suggested, it will break that stage.

This would not be a problem as soon as you resolve https://github.com/ROCmSoftwarePlatform/MIOpen/pull/676#discussion_r568740153

Reply:
[Side note]

> COMgr is targeted for a very specific deliverable...

Well, actually right now we have to maintain two build paths for the HIP backend: offline and COMgr. This means doubling the development and testing efforts. It is highly likely that in the future we will abandon offline build path (and will use only COMgr for the HIP backend). But not now, because COMgr currently has some limitations.

/cc @daniellowell

Reply:
> There is a CI pipeline stage that specifically test COMGR. If I do as suggested, it will break that stage.

I think thats because you are using two different llvm versions.

> COMgr is targeted for a very specific deliverable and can be skipped in the MLIR development.

I think we want to move to using comgr and drop offline support, once comgr's issues are resolved.

Reply:
@pfultz2 
> I think thats because you are using two different llvm versions.

Sorry if I'm not making it clear. By saying break I mean that stage will ended becoming a FATAL_ERROR of "comgr cannot be used with MLIRMIOpen", because libMLIRMIOpen comes as default in the docker image.

> I think we want to move to using comgr and drop offline support, once comgr's issues are resolved.
> It is highly likely that in the future we will abandon offline build path (and will use only COMgr for the HIP backend). 

It is exactly the opposite to what I understand. If this is the case, I will have to implement the missing pieces in COMGR path. Thanks for your clarification guys @pfultz2 @atamazov

Reply:
> ...If this is the case, I will have to implement the missing pieces in COMGR path.

Let's implement MLIR in offline build path for now.

Reply:
This should output an error message if `BUILD_SHARED_LIBS=Off` since we don't provide an imported target and there is potential llvm version conflicts.

Reply:
@pfultz2 I replied this in your thread. Let's continue the discussion in that thread.

@atamazov 
> Let's implement MLIR in offline build path for now.
Sounds good, I can implement the COMGR path in next PR, or whenever necessary.

>     if(MIOPEN_USE_COMGR)
>        message(FATAL_ERROR "comgr cannot be used with MLIRMIOpen")
>     endif()

Would you still suggest me to use this approach in the cmake logic? I still didn't get it such that, in the `Hip Debug COMGR` stage, when both libMLIRMIOpen.a exists and -DMIOPEN_USE_COMGR=On, how do I make the tests in that stage pass? To my understanding, CI pipeline on that stage will just error out in the cmake configuration step and mark it as failed.

Reply:
Validated manually that the `if statement` in CMakeLists.txt will directly fail the `Hip Debug COMGR` stage. I will have to not link in the libMLIRMIOpen.a when COMGR exists. Marking this as resolved.

Reply:
> Would you still suggest me to use this approach in the cmake logic?

Yes.

> I still didn't get it such that, in the Hip Debug COMGR stage, when both libMLIRMIOpen.a exists...

COMGR path implies that we link COMgr (that includes the llvm-based compiler) into the library. IIRC we can't link two different llvms. This opens a question if we can make MLIR frontend (?) upstreamable to the llvm.

Reply:
@atamazov Just to make it clear, linking with COMGR and libmlirmiopen.a won't break any existing MIOpen unit tests. There might be certain very strict conditions where everything is built statically (a situation not tested in MIOpen), that is able to make it break. And I'm not able to see things break just yet. 

In summary, I want to say that the two sets of llvm symbols breakage is not a show-stopper for this PR and can be "fixed" in a later mlir issue [here](https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/80).

Reply:
Resolved. Added `MIOPEN_USE_MLIR` CMAke flag. MLIR will be used only if explicitly asked.